4.4 - Add Controller::viewClasses() and support content negotiation

markstory opened the pull request:
Add the controller and view hook methods for doing content negotiation. I have a few open questions on how this should work.

* [x] Should we be mapping `_ext` to a content type and using in place of the `Accept` header if it exists?
* [x] Given that these new methods have no setter do we need the `get` prefix? Would `Controller::viewClasses()` and `View::contentType()` be a simpler/better method names?

Part of #16113

markstory committed:
Add getContentType to core view classes

This new hook method allows a View class to make the content type that
the view produces a public API. This will enable the new
content-negotiation features to be built.

It will also allow automatic content-type header setup as we can apply
the view's content-type automatically.

markstory committed:
Implement controller methods necessary for content-type negotiation.

This adds view selection into controller and adds simple integration
tests for its integration into `render()`

markstory committed:
Set content type from view class.

ADmad commented:
> Should we be mapping _ext to a content type and using in place of the Accept header if it exists?

Yes please, that's the current behavior and it's reasonable. Can't think of a sane reason why someone would want for e.g.  HTML to be returned for a URL with .json extension.

ADmad commented on src/Controller/Controller.php:
```diff
@@ -757,12 +758,65 @@ public function render(?string $template = null, ?string $layout = null): Respon
         if ($builder->getTemplate() === null) {
             $builder->setTemplate($this->request->getParam('action'));
         }
+        // TODO should this only be done if builder->getViewClass() is null?
```
Yes.

markstory replied:
Done.

markstory committed:
Don't apply content-negotiation if the view has been defined.

If the controller or component has defined the view class to use we
should use it. Content-negotiation should be a fallback.

markstory committed:
Use _ext routing parameter for view selection if available.

This allows routes to force content types based on extensions. This
preserves behavior with RequestHandler and is generally what should
happen.

markstory commented:
>  Given that these new methods have no setter do we need the get prefix? Would Controller::viewClasses() and View::contentType() be a simpler/better method names?

What do folks think about this? I'm leaning towards not having the prefix as `get` implies there is a `set()` but there isn't for this method, nor do I think we should have one.

othercorey commented:
> What do folks think about this? I'm leaning towards not having the prefix as `get` implies there is a `set()` but there isn't for this method, nor do I think we should have one.

Removing get when there is no set is fine with me.

ADmad commented on src/Controller/Controller.php:
```diff
@@ -757,12 +758,79 @@ public function render(?string $template = null, ?string $layout = null): Respon
+        $viewContentType = $view->getContentType();
+        $responseType = $response->getHeaderLine('Content-Type');
+        if ($viewContentType && ($responseType === '' || substr($responseType, 0, 9) === 'text/html')) {
+            $response = $response->withHeader('Content-Type', $viewContentType);
```
IMO it should be the view class' responsibility to set the proper content type. Currently `SerializedView` already do so.

markstory replied:
> IMO it should be the view class' responsibility to set the proper content type. Currently SerializedView already do so.

Where to put that logic though? We could put it the `initialize()` hook or `__construct()`.

ADmad replied:
In `initialize()` sounds good.

markstory replied:
Done now. I also removed the get prefix on getContentType as I don't think it adds anything as there is no set again.

ADmad commented on src/Controller/Controller.php:
```diff
@@ -757,12 +758,79 @@ public function render(?string $template = null, ?string $layout = null): Respon
+        $contentType = new ContentTypeNegotiation();
+        $preferredType = $contentType->preferredType($request, array_keys($typeMap));
+        if (!$preferredType) {
+            return null;
```
I am wondering if we should throw a `NotAcceptableException` exception here. For e.g. if my API only supports JSON and some uses `Accept` to ask for another type then returning a `406 Not Acceptable` response seems reasonable.

markstory replied:
We would have already run the action though. I think an exception could work well if we triggered it before the controller's action was called.

ADmad replied:
True, if there's an exception it should be before the action runs.

ADmad commented on src/Controller/Controller.php:
```diff
@@ -757,12 +758,79 @@ public function render(?string $template = null, ?string $layout = null): Respon
+
+        $typeMap = [];
+        foreach ($possibleViewClasses as $class) {
+            $viewContentType = $class::getContentType();
```
This means custom view classes won't work with 4.4. without adding (or overriding to return proper value) this method. So that breaks BC.

markstory replied:
But controller will also have to implement the `viewClasses()` method. My thinking was that when applications were updated to implement `viewClasses()` they could also update their view classes to support `getContentType()`. For plugin views that are lagging behind developers could subclass and add the method as a shim.

RequestHandler features aren't being moved/removed allowing more cautious applications to have full backwards compatibility.

ADmad replied:
Okay.

markstory committed:
Rename to `viewClasses()`

Shorter name and there is no set

markstory committed:
Update baseline file and formatting.

markstory committed:
Move content type definition to view

* Remove `get` prefix from View::contentType() as there is no set.
* Add content-type setting to View::initialize() in a protected method
  that can be redefined by sub-classes if necessary.
* Update core views to use new methods and logic.

markstory committed:
Ignore deprecated property usage.

markstory commented:
Anything else that needs to be addressed with this?

othercorey commented on src/Controller/Controller.php:
```diff
@@ -757,12 +758,71 @@ public function render(?string $template = null, ?string $layout = null): Respon
+    }
+
+    /**
+     * Get the list of View classes this controller can negotiate with.
```
```suggestion
     * Get the View classes this controller can perform content negotiation with.
```

othercorey commented on src/Controller/Controller.php:
```diff
@@ -757,12 +758,71 @@ public function render(?string $template = null, ?string $layout = null): Respon
+            return null;
+        }
+        if ($this->viewBuilder()->getClassName() !== null) {
+            return null;
```
Why does this return null?

markstory replied:
Either the Controller's action, `beforeRender()` logic or a component set the viewClass so we shouldn't overwrite that decision with content-type negotiation. I've added a comment.

othercorey replied:
Ah, I misread the check as === null.

othercorey commented on src/Controller/Controller.php:
```diff
@@ -757,12 +758,71 @@ public function render(?string $template = null, ?string $layout = null): Respon
+        foreach ($possibleViewClasses as $class) {
+            $viewContentType = $class::contentType();
+            if ($viewContentType && !isset($typeMap[$viewContentType])) {
+                $typeMap[$viewContentType] = $class;
```
Should viewClasses() return this map instead? It seems the user should is already responsible for ordering.

markstory replied:
But then all userland code has to maintain all these method calls. Giving a list of classes lets users write less code and get the same results.

markstory committed:
Add comment explaining behavior.

markstory committed:
Apply suggestions from code review

Co-authored-by: othercorey <corey.taylor.fl@gmail.com>
